server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRecipientRewriteTableMailet.java

```diff
@@ -123,75 +150,159 @@ public void service(Mail mail) throws MessagingException {
      * @param message
      *            the mail message
      * @return a collection of mapped recpient addresses
-     * 
+     *
      * @throws MessagingException
      */
-    protected Collection<MailAddress> handleMappings(Mappings mappings, MailAddress sender, MailAddress recipient, MimeMessage message) throws MessagingException {
-        Iterator<Mapping> i = mappings.iterator();
-        Collection<MailAddress> remoteRecipients = new ArrayList<MailAddress>();
-        Collection<MailAddress> localRecipients = new ArrayList<MailAddress>();
-        while (i.hasNext()) {
-            Mapping rcpt = i.next();
-
-            if (!rcpt.hasDomain()) {
-                // the mapping contains no domain name, use the default domain
-                try {
-                    rcpt = rcpt.appendDomain(domainList.getDefaultDomain());
-                } catch (DomainListException e) {
-                    throw new MessagingException("Unable to access DomainList", e);
+    protected Collection<MailAddress> handleMappings(final Mappings mappings, MailAddress sender, MailAddress recipient, MimeMessage message) throws MessagingException {
+        ImmutableList<Mapping> addressWithoutDomains = getAddressWithDefaultDomain(mappings);
+
+        ImmutableList<Mapping> newMappings = convertToNewMappings(mappings, addressWithoutDomains);
+
+        ImmutableList<MailAddress> mailAddresses = buildMailAddressFromMappingAddress(newMappings);

```

Comment:
A mapping is not a MailAddress. It can be any kind of {Regex, Domain, Error, Address}. In this case, those are addresses so we need to convert to MailAddress.